addons/controllers/packageinstallstatus_controller.go

```diff
@@ -0,0 +1,442 @@
+		return ctrl.Result{}, err
+	}
+
+	// if cluster is marked for deletion, then no reconciliation is needed; just remove the cluster object from the cluster accessor map
+	if !cluster.GetDeletionTimestamp().IsZero() {
+		r.Tracker.DeleteAccessor(req.NamespacedName)
+		return ctrl.Result{}, nil
+	}
+
+	clusterLabels := cluster.GetLabels()
+
+	// make sure the TKR object exists
+	tkrName := clusterLabels[constants.TKRLabelClassyClusters]
+	tkr, err := util.GetTKRByName(ctx, r.Client, tkrName)
+	if err != nil {
+		log.Error(err, "unable to fetch TKR object", "name", tkrName)
+		return ctrl.Result{}, err
+	}
+	// if tkr is not found, should not requeue for the reconciliation
+	if tkr == nil {
+		log.Info("TKR object not found", "name", tkrName)
+		return ctrl.Result{}, nil
+	}
+
+	// determine cluster role
+	_, isManagementCluster := clusterLabels[tkrconstants.ManagementClusterRoleLabel]
+
+	if isManagementCluster {
+		// the cluster is a management cluster
+		clusterRole = clusterRoleManagement
+		c = r.Client
+	} else {
+		// the cluster is a remote workload cluster
+		clusterRole = clusterRoleWorkload
+		remoteClient, err := r.Tracker.GetClient(ctx, clusterapiutil.ObjectKey(cluster))
+		if err != nil {
+			log.Error(err, "error getting remote cluster's client")
+			return ctrl.Result{}, err
+		}
+		c = remoteClient
+		log.Info("successfully got remoteClient")
+		// set watch if not already set. If the watch already exists, it doesn't get re-created (logic in util.RemoteObjectTracker)
+		if err := r.watchPackageInstalls(ctx, cluster, log); err != nil {
+			log.Error(err, "error watching PackageInstalls on target cluster")
+			return ctrl.Result{}, err
+		}
+	}
+
+	if err := r.reconcile(ctx, c, cluster, clusterRole, log); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// reconcile iterates over all core/additional packages and calls reconcileClusterBootstrapStatus() for each.
+// it eventually updates ClusterBootstrapStatus with the condition entries (one for each package) in a single update operation
+// Note: use of the client c (remote or not) depends on whether the corresponding cluster is workload or management which is
+// determined in the Reconcile function
+func (r *PackageInstallStatusReconciler) reconcile(ctx context.Context, c client.Client, cluster *clusterapiv1beta1.Cluster, clusterRole ClusterRole, log logr.Logger) error {
+	clusterBootstrap := &runtanzuv1alpha3.ClusterBootstrap{}
+	// ClusterBootstrap resource exists in the management cluster, that's why we need to use local client instead of remote client for fetching it
+	if err := r.Client.Get(ctx, client.ObjectKeyFromObject(cluster), clusterBootstrap); err != nil {
+		return err
+	}
+	clusterBootstrap = clusterBootstrap.DeepCopy()
+
+	// this shouldn't include kapp ctrl package as it'll get processed separately
+	packages := append([]*runtanzuv1alpha3.ClusterBootstrapPackage{
+		clusterBootstrap.Spec.CNI,
```

Comment:
blocker: we have `clusterBootstrap.Spec.CNIs` now

Reply:
No, I changed that in my other PR which got merged: https://github.com/vmware-tanzu/tanzu-framework/pull/1834